rfc/rfc-66/rfc-66.md

```diff
@@ -0,0 +1,318 @@
+
+In order to unblock the compaction plan scheduling and keep the completeness of the readers, we introduce the completion time for file slice generation:
+
+- Support quick look-up from instant time ➝ completion time, the [HUDI-6539](https://issues.apache.org/jira/browse/HUDI-6539) supports fast completion time queries on archived timeline, based on this, we are able
+to support flexible completion time queries on both active and archived timeline, see [HUDI-6725](https://issues.apache.org/jira/browse/HUDI-6725);
+- New compaction plan scheduling to comply with the completion time, that is: only log files that have smaller completion time(than the compaction start instant time) should be considered
+- New file slice generation strategy, a log file with smaller instant time (than the compaction instant time) but greater completion time should be assigned to a new file slice
+- By combining #2 and #3, in general, we are slicing by comparing the compaction start time with the log files completion time.
+
+<img src="non_serial_compaction.png" alt="drawing" width="400"/>
+
+Assumes we have two log files, with instant time & completion time as [t1, t3] and [t2, t5], at t4, a compaction plan was scheduled,
+the plan does not include file t2_t5_v1.w2, in the reader view, the log file should be assigned to a different file slice than the t4 instant time.
+
+#### Global Monotonically Increasing Timestamp
+
+In order to make the time deterministic among cloud storages, we use a logical time generated by a special **TimeGenerator**, see [HUDI-1623](https://issues.apache.org/jira/browse/HUDI-1623) for details.
+
+#### Assumption On Locks
+In order to get global monotonically increasing time generation, we may introduce lock providers on the instant time generation and the creation of completed metadata file.
+That means, for each instant time/completion time generation request, there is a try-lock action. As a special case, we store the completion time as part of the completed
+metadata file name, the time generation and file creation should have atomicity altogether, they should be guarded under the same lock of regular time generation request.
+
+### The Log File Naming Convention
+
+We use the current instant time instead of the base commit time in the file name,
+so that tasks from different writers can not conflict in file name. We can also parse the file name to fetch the instant time quickly. 
+Finally, the log name is with the following pattern:
+
+```shell
+${uuid}_${delta_commit instant time}.log.${version}_${task_token}
+```
+
+### The Sorting Rules for Log Files from Different Writers
+
+The sorting rules is important because it somehow decides the natual order processing,
+especially when the event time field are the same, and we don't know which payload to choose when combining.
+Here we can keep using the log file name compactor of current codebase, that is:
+
+```sql
+order by delta_commit_time, version_number, write_token
```

Comment:
During read time, Can we use the completion time of the delta-commit to order as this is what we are using for file slice ? 

So completion_time and version_number will give total ordering of the log files.

Reply:
Actually it is not necessary, we have the ordering field to keep the correctness, actually a log file that completes later may be flushed with a smaller start instant time, that means if a file is written first it may be sorted to the latest if it's completion time is the largest.

From my understanding, either we sort by start or end time is okay.